src/libs/io/io.c

```diff
@@ -0,0 +1,620 @@
+}
+
+int elektraIoBindingAddTimer (ElektraIoInterface * binding, ElektraIoTimerOperation * timerOp)
+{
+	if (binding == NULL)
+	{
+		ELEKTRA_LOG_WARNING ("binding cannot be NULL");
+		return -1;
+	}
+	if (timerOp == NULL)
+	{
+		ELEKTRA_LOG_WARNING ("timerOp cannot be NULL");
+		return -1;
+	}
+	if (timerOp->binding != NULL)
+	{
+		ELEKTRA_LOG_WARNING ("operation cannot be assigned to multiple bindings");
+		return -1;
+	}
+	timerOp->binding = binding;
+	return binding->addTimer (binding, timerOp);
+}
+
+int elektraIoBindingUpdateTimer (ElektraIoTimerOperation * timerOp)
+{
+	if (timerOp == NULL)
+	{
+		ELEKTRA_LOG_WARNING ("timerOp cannot be NULL");
+		return -1;
+	}
+	return timerOp->binding->updateTimer (timerOp);
+}
+
+int elektraIoBindingRemoveTimer (ElektraIoTimerOperation * timerOp)
+{
+	if (timerOp == NULL)
+	{
+		ELEKTRA_LOG_WARNING ("timerOp cannot be NULL");
+		return -1;
+	}
+	int result = timerOp->binding->removeTimer (timerOp);
+	timerOp->binding = NULL;
+	return result;
+}
+
+int elektraIoBindingAddIdle (ElektraIoInterface * binding, ElektraIoIdleOperation * idleOp)
+{
+	if (binding == NULL)
+	{
+		ELEKTRA_LOG_WARNING ("binding cannot be NULL");
+		return -1;
+	}
+	if (idleOp == NULL)
+	{
+		ELEKTRA_LOG_WARNING ("idleOp cannot be NULL");
+		return -1;
+	}
+	if (idleOp->binding != NULL)
+	{
+		ELEKTRA_LOG_WARNING ("operation cannot be assigned to multiple bindings");
+		return -1;
+	}
+	idleOp->binding = binding;
+	return binding->addIdle (binding, idleOp);
+}
+
+int elektraIoBindingUpdateIdle (ElektraIoIdleOperation * idleOp)
+{
+	if (idleOp == NULL)
+	{
+		ELEKTRA_LOG_WARNING ("idleOp cannot be NULL");
+		return -1;
+	}
+	return idleOp->binding->updateIdle (idleOp);
+}
+
+int elektraIoBindingRemoveIdle (ElektraIoIdleOperation * idleOp)
+{
+	if (idleOp == NULL)
+	{
+		ELEKTRA_LOG_WARNING ("idleOp cannot be NULL");
+		return -1;
+	}
+	int result = idleOp->binding->removeIdle (idleOp);
+	idleOp->binding = NULL;
+	return result;
+}
+
+int elektraIoBindingCleanup (ElektraIoInterface * binding)
+{
+	if (binding == NULL)
+	{
+		ELEKTRA_LOG_WARNING ("binding cannot be NULL");
+		return -1;
+	}
+	return binding->cleanup (binding);
+}
+
+void * elektraIoBindingGetData (ElektraIoInterface * binding)
+{
+	if (binding == NULL)
+	{
+		ELEKTRA_LOG_WARNING ("operation was NULL");
+		return NULL;
+	}
+
+	return binding->data;
+}
+
+int elektraIoBindingSetData (ElektraIoInterface * binding, void * data)
+{
+	if (binding == NULL)
+	{
+		ELEKTRA_LOG_WARNING ("operation was NULL");
+		return -1;
+	}
+
+	binding->data = data;
+	return 0;
+}
+
+
+// ################################
+// # Constructors for Operations
+// ################################
+
+ElektraIoFdOperation * elektraIoNewFdOperation (int fd, int flags, int enabled, ElektraIoFdCallback callback, void * privateData)
+{
+	if (callback == NULL)
+	{
+		ELEKTRA_LOG_WARNING ("callback cannot be NULL");
+		return NULL;
+	}
+
+	ElektraIoFdOperation * fdOp = elektraMalloc (sizeof (*fdOp));
```

Comment:
again elektraCalloc (isn't memset even useless? You set all fields of the struct, don't you?)

Reply:
Not all, but I now do and got rid of memset.